### DISCUSSION TITLE
onUpdate TPF value isn't accurate and/or onUpdate not always called

### DISCUSSION THREAD

**Discussion by DeathPhoenix22** [AlmasB/FXGL#1222] [General]
Hi all,

I'm facing a problem that's probably deeper in the way TPFs are calculated along with how the onUpdate mechanics is operated.
It seems like time isn't always accurate in the engine, causing all sort of issues.

For instance, when you load a new texture, the TPF isn't properly affected by it. Sometimes, the TPF value is as if loading the texture never affected the time per frame (it took 200 millis load a texture, the onUpdate is not called during that time, but the next onUpdate says -> all good were in 60 fps, tpf = 16.6 millis) This makes the game freeze.

Here is a super basic example of that problem happening in the RTSGame example:

```

...

public class RTSApp extends GameApplication {

    private long lastFrame = System.currentTimeMillis();

...

@Override
    protected void onUpdate(double tpf) {
        long timeBetweenFrames = System.currentTimeMillis() - lastFrame;
        if(timeBetweenFrames - 1 > (int)(tpf * 1000)) {
            System.out.println("WRONG TPF: " + timeBetweenFrames + " VS provided: " + (int)(tpf * 1000));
        }

        lastFrame = System.currentTimeMillis();
        super.onUpdate(tpf);
    }

...

```

In short, it'll printout when the TPF provided is off compared to the last time it was actually called on the component. I noticed that most components will have a frame skipped (2 to 3 times per second).

Thanks.

**Comment by DeathPhoenix22** [comment]
I've found out that TPFs are enforced and doesn't consider slowdowns. Also, if the lag is intense (FPS drop bellow 5 fps), it'll display that 60 FPS were processed.

```
private fun frame(now: Long) {
        if (lastFPSUpdateNanos == 0L) {
            lastFPSUpdateNanos = now
            fpsBuffer2sec = 0
        }

        cpuNanoTime = measureNanoTime {
            runnable(tpf)
        }

        fpsBuffer2sec++

        // if 2 seconds have passed
        if (now - lastFPSUpdateNanos >= 2_000_000_000) {
            lastFPSUpdateNanos = now
            fps = fpsBuffer2sec / 2
            fpsBuffer2sec = 0

            // tweak potentially erroneous reads
            if (fps < 5)
                fps = 60

            // update tpf for the next 2 seconds
            tpf = 1.0 / fps
        }
    }
```

Am I using this engine wrong? By the look of it, it cannot be used if you drop bellow 60 FPS.
